feat: APKAM Enrollment changes

sitaram-kalluri opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- The PR contains changes related to APKAM feature.
   - at_client : 
      - The AtClient package contains changes to expose the enrollment services to the apps such as to fetch the pending enrollments, approve or deny enrollments.
       - Modify SelfKeyEncryption and SelfKeyDecryption to use at_chops instance to fetch the self encryption key key instead of fetching the key from the local secondary.
  - at_client_mobile: 
      - Introduce "enroll", "getSentEnrollmentRequest" and "getFinalEnrollmentStatus" methods in "AtAuthService" class to support APKAM feature.
      - Remove exporting of the implementation classes in at_client_mobile.
      - Updated the port number's docker compose to publish the newly added atSign ports from the container to the host.

- AtClientMobile :
     - To submit a new enrollment, use "enroll" method in the `AtAuthService.` 
     - After submitting an enrollment, to view the enrollment and to check the status, introduced "getSentEnrollmentRequest" and "getFinalEnrollmentStatus" method in `AtAuthService` respectively.
     - Currently, the implementation classes are exported in `at_client_mobile.dart` file. For example `export 'src/auth/at_auth_service_impl.dart'` is being exported. To prevent this, introduce a new class `AtClientMobile` in `at_client_mobile_base.dart` file . This class contains methods to obtain the instances of implementation classes.

- AtClient: 
   - Introduce a new class "EnrollmentService" which exposes methods to fetch the pending enrollment list, approve and deny the enrollments. The "EnrollmentServiceImpl" class implements the "EnrollmentService".
   - The "fetchEnrollmentRequest" forwards the request to server to get the pending enrollment requests. The "PendingEnrollmentRequest" class encapsulates the enrollment details received from the server.
   - The approve and deny requests are forwarded to to at_auth package which fulfills the request.
   - Update the AtClientManager, AtClientSpec, AtClientImpl classes to set the EnrollmentService instance.
   - In the "SelfKeyEncryption" class, use at_chops to fetch the self encryption key instead of fetching from the local secondary stoage.

- Update the existing unit and functional tests to consume the refactoring changes. 

**- How to verify it**
- Tested the changes with the mobile app.

- In AtClientMobile, when an enrollment is approved, to generate atkeys file we use at_file_saver package, with which we are facing issue with latest version. This is worked by @sachins-geekyants in the following PR: https://github.com/atsign-foundation/at_client_sdk/pull/1276

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

murali-shris committed:
feat: APKAM Enrollment changes

sitaram-kalluri committed:
fix: Add dependency_overrides in the pubspec.yaml

sitaram-kalluri committed:
fix: dart analyze changes

sitaram-kalluri committed:
fix: Add ignore for deprecated method and update ports in docker-compose.yaml file

sitaram-kalluri committed:
fix: Run dart formatter

sitaram-kalluri committed:
fix: Add self key encryption to the AtChops instance

gkc commented:
@sitaram-kalluri will review today thank you!

sitaram-kalluri committed:
fix: Add unit tests for enrollment submission

sitaram-kalluri committed:
fix: Add unit tests for getFinalEnrollmentStatus

sitaram-kalluri committed:
fix: Run dart formatter

sachins-geekyants committed:
feat: atkey file saved on enrollment request approval

sachins-geekyants committed:
Merge branch 'apkam_onboarding_changes_with_enrollment_status' of https://github.com/atsign-foundation/at_client_sdk into apkam_onboarding_changes_with_enrollment_status

sitaram-kalluri committed:
fix: Run dart formatter

sitaram-kalluri committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

murali-shris committed:
fix: improved enrollment test, fixed atAuth compile issue in atClient

murali-shris committed:
fix: compile issue in at_client_mobile

murali-shris commented on packages/at_client_mobile/lib/src/auth/at_auth_service_impl.dart:
```diff
@@ -29,4 +127,436 @@ class AtAuthServiceImpl implements AtAuthService {
+    // if enrollmentInfoJsonString is not null, it indicates that there is a pending
+    // enrollment request. So, do not allow another enrollment request.
+    if (enrollmentInfoJsonString != null) {
+      throw InvalidRequestException(
```
can we through AtEnrollmentException here? InvalidRequestException is an exception from secondary server

sitaram-kalluri replied:
The changes are addressed are in https://github.com/atsign-foundation/at_client_sdk/pull/1257/commits/2c3ea9768b9a158ce02499cfe3d836ed3735bae7

gkc committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

murali-shris commented on packages/at_client/lib/src/encryption_service/self_key_encryption.dart:
```diff
@@ -27,48 +24,17 @@ class SelfKeyEncryption implements AtKeyEncryption {
           'Invalid value type found: ${value.runtimeType}. Valid value type is String');
     }
-    // Get AES key for current atSign
-    var selfEncryptionKey =
```
these changes from trunk related to replacing EncryptionUtil methods should not be removed.

sitaram-kalluri replied:
The changes are addressed are in https://github.com/atsign-foundation/at_client_sdk/pull/1257/commits/2c3ea9768b9a158ce02499cfe3d836ed3735bae7

gkc commented:
@sitaram-kalluri This PR is not ready for review. Please ping me once you have addressed @murali-shris's requests and the functional tests are passing

gkc reviewed (changes requested):
Tests are not passing

sitaram-kalluri committed:
fix: Change the exception type to AtEnrollmentException at_auth_service_impl.dart

sitaram-kalluri committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

sitaram-kalluri committed:
fix: Update the dependency_overrides of at_auth to trunk branch

sitaram-kalluri committed:
fix: Dart analyze issues

sitaram-kalluri commented:
> @sitaram-kalluri This PR is not ready for review. Please ping me once you have addressed @murali-shris's requests and the functional tests are passing

@gkc: The review comments are addressed and updated with the latest trunk branch code.

Pending work: Once enrollment is approved, securely store the keys in the keychain manager being worked by @sachins-geekyants. Therefore, the PR is still marked as draft..

murali-shris commented on packages/at_client/pubspec.yaml:
```diff
@@ -40,6 +40,13 @@ dependencies:
   meta: ^1.8.0
   version: ^3.0.2
 
+dependency_overrides:
```
publish at_auth and replace with the published version

sitaram-kalluri replied:
Published at_auth 2.0.0 and updated in pubspec.yaml

sitaram-kalluri committed:
fix: Update at_auth version and remove dependency overrides

murali-shris committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

sitaram-kalluri committed:
fix: update at_auth package version to 2.0.0

sitaram-kalluri committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

gkc commented:
@sitaram-kalluri Have you tested with other things that use the at_client package e.g. at_talk, the various demos in at_demos, noports, etc?

gkc committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

gkc commented:
> > @sitaram-kalluri This PR is not ready for review. Please ping me once you have addressed @murali-shris's requests and the functional tests are passing
> 
> @gkc: The review comments are addressed and updated with the latest trunk branch code.
> 
> Pending work: Once enrollment is approved, securely store the keys in the keychain manager being worked by @sachins-geekyants. Therefore, the PR is still marked as draft..

@sitaram-kalluri  Since the PR is no longer marked as 'draft', can I presume that the Pending work you mention has been completed?

gkc commented on packages/at_client/lib/src/preference/at_client_preference.dart:
```diff
@@ -113,6 +113,9 @@ class AtClientPreference {
   HashingAlgoType hashingAlgoType = HashingAlgoType.sha256;
+
+  // if set to true, the onboarded app will get privilege to approve enrollment requests
+  bool enableEnrollmentDuringOnboard = false;
```
This variable is not well named. The name suggests that you are enabling enrollment during onboarding (which I don't think I understand either) - but the comment says 'the onboarded app will get privilege to approve enrollment requests' ... couple of other questions though
1. This instance variable doesn't appear to be used anywhere in the at_client package. Why is it here at all?
2. What does this actually control and how does it work? Where is the logic that uses this variable?

gkc replied:
I've partially answered my own questions
1. This variable is also declared in the AtOnboardingPreference of the at_onboarding_cli package, which extends AtClientPreference. It should therefore I imagine be removed from AtClientPreference
2. at_onboarding_cli package has explanation and some tests. I would still assert that it is not a well-named variable but we can address that later

gkc replied:
A third question that occurs is - is there logic in the at_auth package which implements the same kind of logic as exists in the at_onboarding_cli package with regards to `enableEnrollmentDuringOnboard` ? If not, why not? is the logic on at_onboarding_cli obsolete?

sitaram-kalluri replied:
Removed "enableEnrollmentDuringOnboard" variable from AtClientPreferences

gkc reviewed (changes requested):
Left some comments and questions but mostly lgtm

sitaram-kalluri committed:
fix: Remove unused variable and run dart format

sitaram-kalluri commented:
@gkc : Completed testing with at_talk and no-ports and works fine. Attaching the log

NoPorts Log : [no-ports.log](https://github.com/atsign-foundation/at_client_sdk/files/14905846/no-ports.log)
at_talk Log : ![at_talk_log](https://github.com/atsign-foundation/at_client_sdk/assets/16645698/6409d549-4686-490c-8cb2-b8ad860b3f28)

sitaram-kalluri committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

gkc reviewed (dismissed):
LGTM. Thank you @sitaram-kalluri

murali-shris committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

gkc commented:
@sitaram-kalluri  Functional tests failing on this test
```
❌ loading test/enrollment_test.dart (failed)
  Failed to load "test/enrollment_test.dart":
  test/enrollment_test.dart:309:29: Error: Method not found: 'FirstEnrollmentRequest'.
      var enrollmentRequest = FirstEnrollmentRequest(
                              ^^^^^^^^^^^^^^^^^^^^^^
```

sitaram-kalluri commented:
> @sitaram-kalluri Functional tests failing on this test
> 
> ```
> ❌ loading test/enrollment_test.dart (failed)
>   Failed to load "test/enrollment_test.dart":
>   test/enrollment_test.dart:309:29: Error: Method not found: 'FirstEnrollmentRequest'.
>       var enrollmentRequest = FirstEnrollmentRequest(
>                               ^^^^^^^^^^^^^^^^^^^^^^
> ```

@gkc : The root cause of the failure is very clear. The "FirstEnrollmentRequest" class is not exported in the at_auth package. We did not export the class because apps will not use the FirstEnrollmentRequest. Apps will invoke the onboard method, which will internally submit the FirstEnrollmentRequest. 

Something that is strange is how did tests pass till now and why the issue occur now? Will debug on this and keep it updated.

sitaram-kalluri committed:
fix: Replace the usage of FirstEnrollmentRequest with executeCommand to submit enrollment request

sitaram-kalluri commented:
> > @sitaram-kalluri Functional tests failing on this test
> > ```
> > ❌ loading test/enrollment_test.dart (failed)
> >   Failed to load "test/enrollment_test.dart":
> >   test/enrollment_test.dart:309:29: Error: Method not found: 'FirstEnrollmentRequest'.
> >       var enrollmentRequest = FirstEnrollmentRequest(
> >                               ^^^^^^^^^^^^^^^^^^^^^^
> > ```
> 
> @gkc : The root cause of the failure is very clear. The "FirstEnrollmentRequest" class is not exported in the at_auth package. We did not export the class because apps will not use the FirstEnrollmentRequest. Apps will invoke the onboard method, which will internally submit the FirstEnrollmentRequest.
> 
> Something that is strange is how did tests pass till now and why the issue occur now? Will debug on this and keep it updated.

@gkc : The issue is fixed. The issue occurs because a particular test is modified to use "FirstEnrollmentRequest". Reverted the change and modified the test as it was before.

gkc committed:
Merge branch 'trunk' into apkam_onboarding_changes_with_enrollment_status

gkc commented:
> > > @sitaram-kalluri Functional tests failing on this test
> 
> > > ```
> 
> > > ❌ loading test/enrollment_test.dart (failed)
> 
> > >   Failed to load "test/enrollment_test.dart":
> 
> > >   test/enrollment_test.dart:309:29: Error: Method not found: 'FirstEnrollmentRequest'.
> 
> > >       var enrollmentRequest = FirstEnrollmentRequest(
> 
> > >                               ^^^^^^^^^^^^^^^^^^^^^^
> 
> > > ```
> 
> > 
> 
> > @gkc : The root cause of the failure is very clear. The "FirstEnrollmentRequest" class is not exported in the at_auth package. We did not export the class because apps will not use the FirstEnrollmentRequest. Apps will invoke the onboard method, which will internally submit the FirstEnrollmentRequest.
> 
> > 
> 
> > Something that is strange is how did tests pass till now and why the issue occur now? Will debug on this and keep it updated.
> 
> 
> 
> @gkc : The issue is fixed. The issue occurs because a particular test is modified to use "FirstEnrollmentRequest". Reverted the change and modified the test as it was before. 

👍 thanks @sitaram-kalluri
